File: src/org/perses/tree/spar/SparTreeBuilder.java
/*
 * Copyright (C) 2018-2020 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */

package org.perses.tree.spar;

import com.google.common.base.Strings;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.RuleNode;
import org.antlr.v4.runtime.tree.TerminalNode;
import org.perses.antlr.GrammarHierarchy;
import org.perses.antlr.RuleHierarchyInfo;
import org.perses.antlr.RuleType;
import org.perses.program.PersesToken;
import org.perses.program.TokenizedProgramFactory;

import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.HashSet;

/** Build a spar-tree from an Antlr {@link ParseTree} node. */
public class SparTreeBuilder {

  private final GrammarHierarchy antlrRuleHierarchy;
  private int nodeIdGenerator;
  private final TokenizedProgramFactory tokenizedProgramFactory;

  public SparTreeBuilder(
      final GrammarHierarchy antlrRuleHierarchy, TokenizedProgramFactory tokenizedProgramFactory) {
    this.antlrRuleHierarchy = antlrRuleHierarchy;
    this.tokenizedProgramFactory = tokenizedProgramFactory;
  }

  private static final boolean areNodeIdsUnique(AbstractSparTreeNode root) {
    final HashSet<Integer> visited = new HashSet<>();
    root.preOrderVisit(
        node -> {
          assert !visited.contains(node.getNodeId()) : "Duplicate node ids: " + node.getNodeId();
          visited.add(node.getNodeId());
          return node.getImmutableChildView();
        });
    return true;
  }

  private static final boolean isEmptyRuleNode(ParseTree node) {
    return isRuleNode(node) && node.getChildCount() == 0;
  }

  private static final boolean isRuleNode(ParseTree node) {
    assert (node instanceof RuleNode) != (node instanceof TerminalNode);
    return node instanceof RuleNode;
  }

  private static final boolean isTokenNode(ParseTree node) {
    assert (node instanceof RuleNode) != (node instanceof TerminalNode);
    return node instanceof TerminalNode;
  }

  public final SparTree build(ParseTree rootParseTree) {
    final HashMap<AbstractSparTreeNode, ParseTree> spar2antlrMap = new HashMap<>();
    final ArrayDeque<AbstractSparTreeNode> stack = new ArrayDeque<>();
    final AbstractSparTreeNode root = createSparTreeNode(rootParseTree);
    assert !spar2antlrMap.containsKey(root);
    spar2antlrMap.put(root, rootParseTree);
    stack.push(root);

    while (stack.size() > 0) {
      final AbstractSparTreeNode node = stack.pop();
      final ParseTree parseTree = spar2antlrMap.get(node);
      assert parseTree != null;
      final int childCount = parseTree.getChildCount();
      for (int i = 0; i < childCount; ++i) {
        final ParseTree child = parseTree.getChild(i);
        if (isEmptyRuleNode(child) || isEOFToken(child)) {
          continue;
        }
        AbstractSparTreeNode sparChild = createSparTreeNode(child);
        assert !spar2antlrMap.containsKey(sparChild);
        spar2antlrMap.put(sparChild, child);
        node.addChild(sparChild);
        stack.push(sparChild);
      }
    }
    assert areNodeIdsUnique(root);
    SparTreeSimplifier.simplify(root);
    return new SparTree(root, tokenizedProgramFactory);
  }

  private final boolean isEOFToken(ParseTree node) {
    return isTokenNode(node) && ((TerminalNode) node).getSymbol().getType() == Token.EOF;
  }

  private AbstractSparTreeNode createSparTreeNode(ParseTree parseTree) {
    final int nodeId = nextNodeId();
    if (isTokenNode(parseTree)) {
      TerminalNode node = (TerminalNode) parseTree;
      final Token symbol = node.getSymbol();
      final int tokenType = symbol.getType();
      final PersesToken persesToken =
          tokenizedProgramFactory.getTokenFactory().getPersesToken(symbol);
      return new LexerRuleSparTreeNode(
          nodeId,
          persesToken,
          RuleType.TOKEN,
          antlrRuleHierarchy.getTokenRule(tokenType));
    }
    final RuleHierarchyInfo rule = antlrRuleHierarchy.getRuleHierarchyInfoForRuleNode(parseTree);
    final RuleType nodeType = computeNodeType(parseTree, rule.getRuleName());
    return new ParserRuleSparTreeNode(nodeId, nodeType, rule);
  }

  private int nextNodeId() {
    return ++nodeIdGenerator;
  }

  private RuleType computeNodeType(ParseTree parseTree, String ruleName) {
    assert !isTokenNode(parseTree);
    assert parseTree instanceof RuleNode : "The class is " + parseTree.getClass();
    assert !Strings.isNullOrEmpty(ruleName) : "The rule name is empty: " + ruleName;

    if (RuleType.isKleeneStar(ruleName)) {
      return RuleType.KLEENE_STAR;
    } else if (RuleType.isKleenePlus(ruleName)) {
      return RuleType.KLEENE_PLUS;
    } else if (RuleType.isOptional(ruleName)) {
      return RuleType.OPTIONAL;
    } else {
      return RuleType.OTHER_RULE;
    }
  }
}

File: src/org/perses/spartree/SparTreeBuilder.kt
/*
 * Copyright (C) 2018-2026 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.spartree

import com.google.common.collect.HashBiMap
import org.antlr.v4.runtime.Token
import org.antlr.v4.runtime.tree.ParseTree
import org.antlr.v4.runtime.tree.RuleNode
import org.antlr.v4.runtime.tree.TerminalNode
import org.perses.antlr.ParseTreeWithParser
import org.perses.grammar.AbstractParserFacade
import org.perses.util.SimpleStack
import org.perses.util.Util.lazyAssert

/** Build a spar-tree from an Antlr [ParseTree] node.  */
class SparTreeBuilder(
  private val sparTreeNodeFactory: SparTreeNodeFactory,
  private val parseTreeWithParser: ParseTreeWithParser,
  val simplifyTree: Boolean,
  val canonicalTokenCountComputer: (() -> Int?),
  val enableNodeActionSetCache: Boolean = false,
) : ISparTreeAntlrTreeMapping {
  private val spar2antlrMap = HashBiMap.create<AbstractSparTreeNode, ParseTree>()
  private var built = false
  val result by lazy {
    build()
  }

  override fun getMappedNodeFor(antlrNode: ParseTree): AbstractSparTreeNode? {
    check(built) { "The spartree must be built" }
    return spar2antlrMap.inverse()[antlrNode]
  }

  override fun getMappedNodeFor(sparTreeNode: AbstractSparTreeNode): ParseTree? {
    check(built) { "The spartree must be built" }
    return spar2antlrMap[sparTreeNode]
  }

  private fun build(): SparTree {
    check(!built) { "The spartree is already built." }
    built = true
    val rootParseTree = parseTreeWithParser.tree
    val spar2antlrMap = spar2antlrMap
    val stack = SimpleStack<AbstractSparTreeNode>()
    val root = createSparTreeNode(rootParseTree)
    lazyAssert { !spar2antlrMap.containsKey(root) }
    spar2antlrMap[root] = rootParseTree
    stack.add(root)
    while (stack.isNotEmpty()) {
      val node = stack.remove()
      val parseTree = spar2antlrMap[node]!!
      val childCount = parseTree.childCount
      for (i in 0 until childCount) {
        val child = parseTree.getChild(i)
        if (isEmptyRuleNode(child) || isEOFToken(child) || isFabricatedTokenNode(child)) {
          continue
        }
        val sparChild = createSparTreeNode(child)
        lazyAssert { !spar2antlrMap.containsKey(sparChild) }
        spar2antlrMap[sparChild] = child
        node.addChild(sparChild, AbstractNodePayload.SinglePayload(sparChild.antlrRule))
        stack.add(sparChild)
      }
    }
    lazyAssert { areNodeIdsUnique(root) }
    if (simplifyTree) {
      SparTreeSimplifier.simplify(root)
    }
    return SparTree(
      realRoot =
        if (root.isPermanentlyDeleted) {
          null
        } else {
          root
        },
      sparTreeNodeFactory = sparTreeNodeFactory,
      initialCanonicalTokenCount = canonicalTokenCountComputer?.invoke(),
      enableNodeActionSetCache = enableNodeActionSetCache,
      hasSyntaxErrors = parseTreeWithParser.hasError,
    )
  }

  private fun isEOFToken(node: ParseTree): Boolean =
    isTokenNode(node) && (node as TerminalNode).symbol.type == Token.EOF

  // Fabricated tokens (only ever present in a tolerantly-parsed tree) carry synthetic text such as
  // "<missing ';'>" and must not become leaves, or they would corrupt the reconstructed program.
  private fun isFabricatedTokenNode(node: ParseTree): Boolean =
    isTokenNode(node) && AbstractParserFacade.isFabricatedToken((node as TerminalNode).symbol)

  private fun createSparTreeNode(parseTree: ParseTree): AbstractSparTreeNode {
    if (isTokenNode(parseTree)) {
      return sparTreeNodeFactory.createLexerRuleSparTreeNode(
        parseTree as TerminalNode,
      )
    }
    return sparTreeNodeFactory.createParserRuleSparTreeNode(
      parseTree as RuleNode,
      parseTreeWithParser.parser,
    )
  }

  companion object {
    fun isEmptyRuleNode(node: ParseTree): Boolean = isRuleNode(node) && node.childCount == 0

    fun isRuleNode(node: ParseTree): Boolean {
      lazyAssert { node is RuleNode != node is TerminalNode }
      return node is RuleNode
    }

    fun isTokenNode(node: ParseTree): Boolean {
      lazyAssert { node is RuleNode != node is TerminalNode }
      return node is TerminalNode
    }

    fun isEOFToken(node: ParseTree): Boolean =
      isTokenNode(node) && (node as TerminalNode).symbol.type == Token.EOF

    private fun areNodeIdsUnique(root: AbstractSparTreeNode): Boolean {
      val visited = HashSet<Int>()
      root.preOrderVisit { node: AbstractSparTreeNode ->
        lazyAssert({ !visited.contains(node.nodeId) }) { "Duplicate node ids: " + node.nodeId }
        visited.add(node.nodeId)
        node.immutableChildView
      }
      return true
    }

    fun createSparTree(
      sparTreeNodeFactory: SparTreeNodeFactory,
      parseTreeWithParser: ParseTreeWithParser,
      canonicalTokenCountComputer: () -> Int,
      enableNodeActionSetCache: Boolean = false,
    ): SparTree =
      SparTreeBuilder(
        sparTreeNodeFactory,
        parseTreeWithParser,
        simplifyTree = true,
        canonicalTokenCountComputer,
        enableNodeActionSetCache = enableNodeActionSetCache,
      ).result

    fun createSparTreeNode(
      sparTreeNodeFactory: SparTreeNodeFactory,
      parseTreeWithParser: ParseTreeWithParser,
      canonicalTokenCountComputer: () -> Int,
    ) = createSparTree(
      sparTreeNodeFactory,
      parseTreeWithParser,
      canonicalTokenCountComputer,
    ).detachRootFromTree()
  }
}
